src/connectors/impls/gpubsub/gsub.rs

```diff
@@ -0,0 +1,284 @@
+use crate::connectors::google::AuthInterceptor;
+use crate::connectors::prelude::*;
+use async_std::channel::{Receiver, Sender};
+use async_std::prelude::FutureExt;
+use async_std::task::JoinHandle;
+use dashmap::DashMap;
+use googapis::google::pubsub::v1::subscriber_client::SubscriberClient;
+use googapis::google::pubsub::v1::{
+    AcknowledgeRequest, PubsubMessage, PullRequest, ReceivedMessage,
+};
+use gouth::Token;
+use serde::Deserialize;
+use std::fmt::Debug;
+use std::sync::Arc;
+use std::time::Duration;
+use tonic::codegen::InterceptedService;
+use tonic::transport::{Certificate, Channel, ClientTlsConfig};
+use tonic::Status;
+use tremor_pipeline::ConfigImpl;
+
+#[derive(Deserialize, Clone)]
+struct Config {
+    pub connect_timeout: u64,
+    pub subscription_id: String,
+}
+impl ConfigImpl for Config {}
+
+#[derive(Debug, Default)]
+pub(crate) struct Builder {}
+
+#[async_trait::async_trait]
+impl ConnectorBuilder for Builder {
+    fn connector_type(&self) -> ConnectorType {
+        "gsub".into()
+    }
+
+    async fn build(&self, alias: &str, raw_config: &ConnectorConfig) -> Result<Box<dyn Connector>> {
+        if let Some(raw) = &raw_config.config {
+            let config = Config::new(raw)?;
+            Ok(Box::new(GSub { config }))
+        } else {
+            Err(ErrorKind::MissingConfiguration(alias.to_string()).into())
+        }
+    }
+}
+
+struct GSub {
+    config: Config,
+}
+
+type PubSubClient = SubscriberClient<InterceptedService<Channel, AuthInterceptor>>;
+type AsyncTaskMessage = Result<(u64, Vec<u8>)>;
+
+struct GSubSource {
+    config: Config,
+    client: Option<PubSubClient>,
+    receiver: Option<Receiver<AsyncTaskMessage>>,
+    ack_ids: Arc<DashMap<u64, String>>,
+    task_handle: Option<JoinHandle<()>>,
+}
+
+impl GSubSource {
+    pub fn new(config: Config) -> Self {
+        GSubSource {
+            config,
+            client: None,
+            receiver: None,
+            ack_ids: Arc::new(DashMap::new()),
+            task_handle: None,
+        }
+    }
+}
+
+async fn consumer_task(
+    mut client: PubSubClient,
+    sender: Sender<AsyncTaskMessage>,
+    ack_ids: Arc<DashMap<u64, String>>,
+    subscription_id: String,
+    request_timeout: Duration,
+) {
+    let mut ack_counter = 0;
+
+    loop {
+        let response_with_potential_timeout = client
+            .pull(PullRequest {
```

Comment:
The streaming call would maybe give us some savings, as we'd send less requests, but I'm not convinced it's worth the complexity (we'd likely need another async task to handle producing the request stream). There will be only one connection anyway, the only saving here is a bit less requests (but we still need to send requests with ACKs, so it's not that much)

Reply:
streaming also reduces latency which can be a significant impact, plus given the amount of messages gpubsub is likely to see savings do make some sense.

Reply:
Agreed, might not be worth it.

Reply:
Just saw @Licenser commenting. Yeah, it might be worth it checking the diff, especially with high volumes.

Making this work would require to have 2 channels, one for acks into the background task and one from the background task to the source. This might help get rid of the `dashmap` as sync point and move the mapping from pull_id to ack_id into the background task.

But i am also happy with creating a separate issue for this enhancement and tackle it later. We are going to fetch `QSIZE` messages at a time, so latency might not be suffering too much.